zookeeper-client/zookeeper-client-c/src/zookeeper.c

```diff
@@ -3495,6 +3495,51 @@ static int add_multi_completion(zhandle_t *zh, int xid, void_completion_t dc,
     return add_completion(zh, xid, COMPLETION_MULTI, dc, data, 0,0, clist);
 }
 
+/**
+ * After sending the close request, we are waiting for a given millisecs for
+ * getting the answer and/or for the socket to be closed by the server.
+ *
+ * This function should not be called while we still want to process
+ * any response from the server. It must be called after adaptor_finish called,
+ * in order not to mess with the I/O receiver thread in multi-threaded mode.
+ */
+int wait_for_session_to_be_closed(zhandle_t *zh, int timeout_ms)
+{
+    int ret = 0;
+#ifndef WIN32
+    struct pollfd fd_s[1];
+#else
+    fd_set rfds;
+    struct timeval waittime = {timeout_ms / 1000, (timeout_ms % 1000) * 1000000};
+#endif
+
+    if (zh == NULL) {
+        return ZBADARGUMENTS;
+    }
+
+#ifndef WIN32
+    fd_s[0].fd = zh->fd->sock;
+    fd_s[0].events = POLLIN;
+    ret = poll(fd_s, 1, timeout_ms);
+#else
+    FD_ZERO(&rfds);
+    FD_SET(zh->fd->sock , &rfds);
+    ret = select(0, &rfds, NULL, NULL, &waittime);
+#endif
+
+    if (ret == 0){
+        LOG_WARN(LOGCALLBACK(zh), "Timed out (%dms) during waiting for server's reply after sending a close request, sessionId=%#llx\n",
+            timeout_ms, zh->client_id.client_id);
+    } else if (ret < 0) {
+        LOG_WARN(LOGCALLBACK(zh), "System error (%d) happened while waiting for server's reply, sessionId=%#llx\n",
+            ret, zh->client_id.client_id);
+    } else {
+        // do nothing
```

Comment:
Should we read the data that server sent us before closing?
I will check what does serve send as response to ZOO_CLOSE_OP(i.e OpCode.closeSession from server side).

Reply:
In general I am not sure if it matters at this stage. I mean we will not behave differently regardless what the server responds. (I think we will actually get back an ACK in https://github.com/apache/zookeeper/blob/e4758ba70b35dd38ff0bba9e534790391f53b050/zookeeper-server/src/main/java/org/apache/zookeeper/server/FinalRequestProcessor.java#L330 ). I can verify it by wireshark if needed. Do you think it matters if we just skip the processing of the response?

Reply:
You are right Mate. It doesn't matter in the context of this bug. I was just trying to understand. You can please go ahead with your changes.